### ISSUE TITLE
FreeBSD service file not functional

### ISSUE THREAD

**Issue by yujiri8** [yggdrasil-network/yggdrasil-go#691]
Yggdrasil version 0.3.14, FreeBSD 12.1-RELEASE-p2.

I'm able to get my Digital Ocean droplet running an Yggdrasil node by running the commands directly, but not with the service file. After experimenting, I've identified several issues with it.

The current service file `contrib/freebsd/yggdrasil` wraps the binary by creating the network interface before starting `yggdrasil` and destroying it when it exits. First of all, at least the interface creation is unnecessary because the binary by itself creates the interface on startup. (Though it doesn't seem to destroy it.)

Second, the way it gets the interface name is extremely kludgy: it `grep`s the config file for a pattern including `/dev/tap`. This is also confusing because in the generated HJSON yggdrasil.conf, the comment on the IfName variable says it's a TUN adapter, but the service script can only find it if it's named as */dev/tap* instead of /dev/tun. This led to me learning [the difference between these things](https://en.wikipedia.org/wiki/TUN/TAP), and it sounds like TUN is the correct one.

So I think the solution is to have the binary manage that and not the service file.

Another issue I see is that the command args to the `daemon` utility include `-f`, which redirects log output to /dev/null. I don't why that's in there (especially since it performs stream redirection to `/var/log/yggdrasil.stdout.log` and `/var/log/yggdrasil.stderr.log`). Explains why the log files were always empty. If I take the `-f` flag out, the log files work properly. (There's another log issue, namely that it seems to be block-buffered so I only see the log output from startup appear after `service yggdrasil stop`, but that might be something I'm doing wrong somehow.)

Another thing the service file does that's probably unintentional is use write mode instead of append mode for the log files, meaning log output from previous runs is wiped if you restart the service.

After making the changes I suggest to the service file, I can get it to start as a service correctly (but still not remove the interface on shutdown).

**Comment by neilalexander** [comment]
The stuff that's in `contrib` is quite possibly out of date or broken. The more up-to-date sources for the FreeBSD packages are at https://github.com/yggdrasil-network/yggdrasil-package-freebsd although they also haven't been updated for the last release or two.

I've rather been hoping that someone who really groks FreeBSD will come along and take over maintainership of the package. I threw it together but don't actually use FreeBSD.

If you are willing, please PR fixes to the above repo. It should build in [Poudriere](https://www.freebsd.org/doc/en_US.ISO8859-1/books/porters-handbook/testing-poudriere.html) for testing.

**Comment by yujiri8** [comment]
I see. I had tried installing yggdrasil from the package repository, but it included an old version (0.3.11). I considered making a PR to update the FreeBSD port, but ended up deciding it wasn't very productive, since to work on Yggdrasil that way I would have to update the port every time there was a bugfix. And the FreeBSD committers are usually very slow to respond to things like that.

I guess it has to be done though. I'll do a little more investigation on this and PR to that repo.

If that's the direction this is going, what will happen to `contrib/freebsd`? Should it be removed?

Also, for this to be fully solved, the binary would need to take on the responsibility of destroying the interface on shutdown. Otherwise the service file is stuck grepping through the config file to try to find it. I guess I could make the service file destroy the interface on shutdown but not create it on startup as a temporary solution.

**Comment by yujiri8** [comment]
Update: I think I have a working port update that solves these issues except for destroying the interface on shutdown. That change has to be made in the main program before I can know what to do here (assuming you agree about that being the right approach). If yes, I could be the one to do that.

**Comment by yujiri8** [comment]
Hey. I haven't revisited this in so long because I was waiting on you to make the binary handle the interface destruction. It looks like something changed since then, now it seems to destroy the interface on shutdown unless it's running under `daemon`, which it is on FreeBSD. I haven't figured out the cause yet.

**Comment by yujiri8** [comment]
I've tracked it a little further. Yggdrasil is not receiving the TERM signal when it's delivered to `daemon`. It works if daemon runs without the `-P` flag which is what makes it fork, but then it doesn't have a PID file and `service yggdrasil status` would always report it as not running.

It must be something Yggdrasil is doing, because when I run this:
```go
package main

import (
	"io/ioutil"
	"os"
	"os/signal"
	"syscall"
)

func main() {
	var ch = make(chan os.Signal, 1)
	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
	for {
		select {
		case s := <-ch:
			ioutil.WriteFile("go-log", []byte(s.String()), 0644)
			os.Exit(0)
		}
	}
}
```
Using the same daemon command, `daemon -P go.pid ./test`, it's able to execute the case when I TERM the daemon process. But for Yggdrasil I see nothing even if I put before `goto exit` in that case.

**Comment by yujiri8** [comment]
I finally figured out this shutdown thing. It's because of SIGPIPE! Yggdrasil logs stuff when it exits, but `daemon` has closed the pipes, so Yggdrasil gets SIGPIPE and dies before it can remove the interface. If it doesn't try to write anything to stdout or stderr, the problem goes away.

Apparently the FreeBSD `daemon` doesn't have a way to give the child the stdout and stderr directly. But using Yggdrasil's own `-logto` option can get around it, since the log output won't be sent through `daemon`.

I think I can finish that PR on the FreeBSD port repository now.
